eclipse-che/che-devfile-registry
# feat: add devfile schema validation to the generator

### What does this PR do?
Integrates devfile schema validator into the devworkspace generator tool. The validator checks devfile before generating a dev workspace. In case if the devfile is not valid, it'll show validation problems.

The validator takes into the account next devfile's schemes: **2.0.0**, **2.1.0**, **2.2.0** and  **2.2.1-alpha**

### Screenshot/screencast of this PR


![screenshot-nimbusweb me-2023 11 09-13_10_19](https://github.com/eclipse-che/che-devfile-registry/assets/1271546/dd9772e2-6818-4e80-94c0-01bccc54dea1)


### What issues does this PR fix or reference?

https://github.com/eclipse/che/issues/22594

### How to test this PR?

1. By using a command line:
```bash
npx @svor/che-devworkspace-generator@7.77.0-7-next \
        --devfile-url:https://github.com/svor/web-nodejs-sample/tree/no-valid-devfile \
        --plugin-registry-url:https://che-plugin-registry-main.surge.sh/v3/ \
        --editor-entry:che-incubator/che-code/latest \
        --output-file:/tmp/devworkspace-che-code-latest.yaml
```
The generator should fail with a message:
```
Validating devfile
stack=Error: Devfile schema validation failed. Error: 0: instance.commands[0].exec requires property "commandLine"
1: instance.commands[0].exec requires property "component"
```

2. Uing the User dashboard. 

For this need to deploy Che with custom dashboard image: quay.io/vsvydenk/che-dashboard:new-generator. This image uses a new version of the devworkspace generator and provides a new sample with invalid devfile.

To test changes just try to create a workspace from **Empty Workspace Invalid Command** sample. This sample uses invalid devfile:
```yaml
schemaVersion: 2.2.0
metadata:
  name: empty
  namespace: svor-che
components:
  - name: universal-developer-image
    container:
      image: quay.io/devfile/universal-developer-image:ubi8-latest
commands:
  - id: command-id
    exec:
      label: "2. Configure Che to use the Custom Devfile Registry"
      # invalid devfile: no componnet
      # component: builder
      # invalid devfile: no commandLine
      # commandLine: ls
``` 
As a result you should see 

![screenshot-nimbusweb me-2023 11 09-13_10_19](https://github.com/eclipse-che/che-devfile-registry/assets/1271546/ad374d92-3229-4aa6-8991-0ce64188c521)


### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)


### Reviewers

Reviewers, please comment how you tested the PR when approving it.
